@here/harp-mapview/lib/text/TextElementsRenderer.ts

```diff
@@ -1411,108 +1315,80 @@ export class TextElementsRenderer {
+     *
+     * @param time Current time for animations.
+     * @param frameNumber Integer number incremented every frame.
+     */
+    private renderTextElements(time: number, frameNumber: number) {
+        const mapViewState: MapViewState = {
+            cameraIsMoving: this.m_mapView.cameraIsMoving,
+            maxVisibilityDist: this.m_mapView.viewRanges.maximum,
+            zoomLevel: this.m_mapView.zoomLevel,
+            frameNumber,
+            time,
+            numRenderedTextElements: 0,
+            fadeAnimationRunning: false
+        };
 
-                if (textElement.tileCenter === undefined) {
-                    textElement.tileCenter = new THREE.Vector3(
-                        tile.center.x + worldOffsetX,
-                        tile.center.y,
-                        tile.center.z
-                    );
-                } else {
-                    textElement.tileCenter.set(
-                        tile.center.x + worldOffsetX,
-                        tile.center.y,
-                        tile.center.z
-                    );
-                }
+        const renderStartTime =
+            this.overloaded && this.m_mapView.isDynamicFrame ? PerformanceTimer.now() : undefined;
 
-                // If the distance is greater than allowed, skip it.
-                const textDistance = this.updateViewDistance(
-                    this.m_mapView.worldCenter,
-                    textElement
-                );
-                if (this.m_mapView.projection.type === ProjectionType.Spherical) {
-                    tempPoiPosition.copy(textElement.position).add(textElement.tileCenter!);
-                    tempPoiPosition.normalize();
-                    const cameraDir = new THREE.Vector3();
-                    this.m_mapView.camera.getWorldDirection(cameraDir);
-                    if (
-                        tempPoiPosition.dot(cameraDir) < -0.6 &&
-                        textDistance !== undefined &&
-                        textDistance <= maxDistance
-                    ) {
-                        tile.placedTextElements.add(textElement);
-                    }
-                } else if (textDistance !== undefined && textDistance <= maxDistance) {
-                    tile.placedTextElements.add(textElement);
-                }
-            }
+        if (renderStats) {
+            renderStats.clear();
+        }
+
+        if (this.m_textRenderers.length === 0) {
+            logger.warn("No text renderers initialized.");
+            return;
+        }
+        if (this.m_textElementStateCache.size === 0) {
+            logger.debug("Text element cache empty.");
+            return;
+        }
+
+        // TODO: HARP-7648. Potential performance improvement. Render persistent labels + rejected
+        // candidates from previous frame if there's been no placement in this one.
+        this.renderPass(Pass.PersistentLabels, mapViewState, renderStartTime);
```

Comment:
correct me if I'm wrong, but this is not quite in line with harp-native, there we iterate over all groups, update all label groups of a given priority that are persistent, then when the priority of the group changes increases, we place all new labels of the previous label groups

Reply:
true, this is wrong. I'll fix it.

Reply:
that's true. I'll fix it.